### ISSUE TITLE
Unnecessary redeployments?

### ISSUE THREAD

**Issue by aprilmintacpineda** [aws/aws-sam-cli#2558]
### Describe your idea/feature/enhancement

I'm just wondering why AWS SAM CLI builds and deploys every function on every deployments, instead of just deploying what changed, I edit one function and deploy and the whole app gets deployed. This makes deployment really slow and it gets slower as you add more functions to it.

For context:

**How I deploy**

I use github actions, here's my yml file:

```yml

name: AWS SAM deployment

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the dev branch
on:
  push:
    branches: [dev]
  pull_request:
    branches: [dev]

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  build:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest
    continue-on-error: false
    env:
      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@master

      - name: Versions
        run: |
          echo "SAM: $(sam --version)"
          echo "Yarn: $(yarn --version)"
          echo "Node: $(node --version)"

      - name: Install dependencies
        run: yarn

      - name: Transpile JS
        run: |
          NODE_ENV=production yarn build
          cd build/dependencies/nodejs && yarn
          rm -rf build/dependencies/nodejs/package.json build/dependencies/nodejs/yarn.lock

      - name: Sam build & deploy
        run: |
          sam build
          sam deploy --stack-name my-app \
            --s3-prefix my-app \
            --parameter-overrides Stage=dev AppPackageName=my-app
```

Besides this, I also can see on my local when I run `sam build` that it includes basically the rest of the files for the function. In my case my project looks like this:

```
/ root
| - package.json
| - yarn.lock
| - node_modules
| - README.md
| - src <-- source folder
| - build <-- result of src built using babel
|   | - dependencies <-- this is a layer
|   | - functions <-- this has all the functions I have in the app
|   |   | - auth <-- all functions for logged in users, it has JS files
|   |   | - events <-- all functions relating to events
|   |   |   | - invoke <-- all invokable functions, has js files in it.
|   |   |   | - s3 <-- all s3 related events, has js files in it
|   |   | - guest <-- all functions for not logged in users, it has JS files
```

The build will look like:

```
/.aws-sam
| -build.toml
| - build
|   | - login
|   |   | - package.json
|   |   | - yarn.lock
|   |   | - README.md
|   |   | - node_modules <-- node_modules from the root
|   |   | - build <--- same as build above, including all files and folders
```

So why would the login need the whole build itself when it just needs `build/functions/guest/login.js`.

One theory I have for why it works like this is because SAM can't identify if you modularized your files, so in case of the login, maybe inside is a line like this: `const someFile = require('../../somefile);` which is understandable in that case, but I think in most cases this wouldn't happen as we would most likely put our files on a lambda layer like what I am doing, the `dependencies` is a lambda layer that has all the utility functions. Is there a way that we can avoid this kind of behavior?

### Proposal

AWS SAM CLI should just redeploy the functions that changed.

Things to consider:
1. Will this require any updates to the [SAM Spec](https://github.com/awslabs/serverless-application-model)

None that I can see.

### Additional Details

N/A

**Comment by ssenchenko** [comment]
Thanks for the feature request! We have a [roadmap](https://github.com/aws/aws-sam-cli/wiki/SAM-CLI-Roadmap) here, but we also try to take in popular feature requests based on 👍🏽 from the community. We will be adding this to our intake for prioritization.

**Comment by mndeveci** [comment]
Thanks for the feedback @aprilmintacpineda

For `sam build` you can use `--cached` parameter which won't build application if files in `CodeUri` folder haven't changed.

For `sam deploy` and `sam package`, they already have a way to validate if the packaged functions (zip files) have changed or not. In that sense if your zip file is same with what is been deployed before, this should be skipped. Can you check if files are uploaded every time when you do deploy even though the code haven't changed?

Thanks!

**Comment by aprilmintacpineda** [comment]
> Thanks for the feedback @aprilmintacpineda
> 
> For `sam build` you can use `--cached` parameter which won't build application if files in `CodeUri` folder haven't changed.
> 
> For `sam deploy` and `sam package`, they already have a way to validate if the packaged functions (zip files) have changed or not. In that sense if your zip file is same with what is been deployed before, this should be skipped. Can you check if files are uploaded every time when you do deploy even though the code haven't changed?
> 
> Thanks!

That also means that I need to save the cache folder, what's the path for the caches for sam?

**Comment by mndeveci** [comment]
Default value is `.aws-sam/cache` but you can override it by providing `--cache-dir` option.

**Comment by jfuss** [comment]
Issue seems to be answered and no further action needed. If you are still experiencing an issue, feel free to re-open or create a new issue.
